### PR TITLE
client-go: export NewDebuggingRoundTripper function and DebugLevel

### PR DESCRIPTION
**Pull request by atosatto**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

`debuggingRoundTripper` is a useful tool to debug Kubernetes API requests and their timing.

Unfortunately it can only be used via the `DebugWrappers` function, which adjusts the level of debugging information logged by the roundTripper based on the configured `klog` verbosity.

While `DebugWrappers` definitely works for clients using `klog` (like `kubectl`), this is currently hard to be used from controllers using `controller-runtime`, as by default the library uses `github.com/go-logr/logr` for logging.

In this PR we change the visibility of `newDebuggingRoundTripper` and `debugLevel` in order to be able to use directly the `debugRoundTripper` without the `DebugWrappers` function.

In particular, the changes proposed in this PR allow to use the `debuggingRoundTripper` to intercept Kubernetes API requests performed by `controller-runtime`  using the following sample code

```go
import (
        ctrl "sigs.k8s.io/controller-runtime"
)

func init() {
	ctrl.SetLogger(zap.New())
}

func main() {
        // wrap the http transport used by the Kubernetes client
	restConfig, err := ctrl.GetConfig()
	checkError(setupLog, err, "unable to get kubernetes client config")

        restConfig.Wrap(func(rt http.RoundTripper) http.RoundTripper {
		return transport.NewDebuggingRoundTripper(rt, transport.DebugJustURL)
	})

        ...
}
```

**Does this PR introduce a user-facing change?**:

```release-note
Export NewDebuggingRoundTripper function and DebugLevel options in the k8s.io/client-go/transport package.
```

### CODE REVIEW

**Comment by leilajal** [comment]
cc @lavalamp

**Comment by atosatto** [comment]
@lavalamp adding here some more information on my PR.

Currently, the one way we can build a  `debuggingRoundTripper` is using the `DebugWrappers` function:

```
// DebugWrappers wraps a round tripper and logs based on the current log level.
func DebugWrappers(rt http.RoundTripper) http.RoundTripper {
	switch {
	case bool(klog.V(9).Enabled()):
		rt = newDebuggingRoundTripper(rt, debugCurlCommand, debugURLTiming, debugResponseHeaders)
	case bool(klog.V(8).Enabled()):
		rt = newDebuggingRoundTripper(rt, debugJustURL, debugRequestHeaders, debugResponseStatus, debugResponseHeaders)
	case bool(klog.V(7).Enabled()):
		rt = newDebuggingRoundTripper(rt, debugJustURL, debugRequestHeaders, debugResponseStatus)
	case bool(klog.V(6).Enabled()):
		rt = newDebuggingRoundTripper(rt, debugURLTiming)
	}

	return rt
}
```

Unfortunately this is relying on `klog`'s verbosity to customise the level of informations exposed by the `debuggingRoundTripper` and there is no way for consumers of the `client-go` API to customise the level of information printed to the output by the `debuggingRoundTripper`.

In this PR I proposing to change the visibility of `newDebuggingRoundTripper ` (and `debugLevels`) to allow `client-go`'s API consumers (in my case I am doing it in a `controller-runtime` controller) to create  `debuggingRoundTripper` with custom `debugLevels`.

File: staging/src/k8s.io/client-go/transport/round_trippers.go [hunk 1]
```diff
@@ -353,25 +353,35 @@ func (r *requestInfo) toCurl() string {
...
 )
 
-func newDebuggingRoundTripper(rt http.RoundTripper, levels ...debugLevel) *debuggingRoundTripper {
+// NewDebuggingRoundTripper allows to display in the logs output debug information
+// on the API requests performed by the client.
+func NewDebuggingRoundTripper(rt http.RoundTripper, levels ...DebugLevel) *debuggingRoundTripper {
 	drt := &debuggingRoundTripper{
 		delegatedRoundTripper: rt,
-		levels:                make(map[debugLevel]bool, len(levels)),
+		levels: 			   make(map[DebugLevel]bool, len(levels)),
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/transport/round_trippers.go:384]: nit: tabs -> spaces

**Review by lavalamp — COMMENTED** [review]
I think I'm pretty OK with this since it is just making something public. But did we consider letting the user also pass a log alternative in? So it doesn't have to use the global one? I guess that could be a separate change if we wanted to do that.

**Comment by atosatto** [comment]
> I think I'm pretty OK with this since it is just making something public. But did we consider letting the user also pass a log alternative in? So it doesn't have to use the global one? I guess that could be a separate change if we wanted to do that.

I did but I am afraid this would be a bigger change (perhaps with more contention) as it would require us to switch to the `logr.Logger` interface and default to `klogr` to guarantee backward compatibility with the current behaviour.

I am happy to give it a try in a separated PR as this would definitely help in the scenario where the API  is used in `controller-runtime`'s controllers.

File: staging/src/k8s.io/client-go/transport/round_trippers.go (hunk 1 above)
Comment (atosatto) [reply staging/src/k8s.io/client-go/transport/round_trippers.go]: Fixed in 9cb5580.

**Comment by lavalamp** [comment]
OK. Let's just take this and wait for the structured logging folks to work their way through the system.

/lgtm
/approve

File: staging/src/k8s.io/client-go/go.mod [hunk 2]
```diff
@@ -28,6 +28,7 @@ require (
 	golang.org/x/time v0.0.0-20200630173020-3af7569d3a1e
 	k8s.io/api v0.0.0
 	k8s.io/apimachinery v0.0.0
+	k8s.io/klog v1.0.0
 	k8s.io/klog/v2 v2.4.0
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/go.mod:32]: Doesn't seem right to have both v1 and v2 in this list?

Comment (atosatto) [reply staging/src/k8s.io/client-go/go.mod]: Indeed! At It was also unexpected.
It should be fixed in 961fa25.

**Comment by atosatto** [comment]
@lavalamp it looks like the PR is now passing all the tests. Would you mind checking again whether it is good to be merged? Thanks!

**Pull request merged**